fdbserver/TagPartitionedLogSystem.actor.cpp

```diff
@@ -1998,9 +1998,17 @@ Optional<std::tuple<Version, Version, std::vector<TLogLockResult>>> TagPartition
 		int safe_range_begin = logSet->tLogWriteAntiQuorum;
 		int new_safe_range_begin = std::min(logSet->tLogWriteAntiQuorum, (int)(results.size() - 1));
 		int safe_range_end = std::max(logSet->tLogReplicationFactor - absent, 1);
-
-		if (!lastEnd.present() || ((safe_range_end > 0) && (safe_range_end - 1 < results.size()) &&
-		                           results[safe_range_end - 1].end < lastEnd.get())) {
+		// The index (in "results" vector) of the recovery version that we will use in the check below
+		// to decide whether to restart recovery not. In "main" we use the version at index
+		// "(safe_range_end - 1)" - this is to minimize the chances of restarting the current recovery
+		// process. With "version vector" we use the version at index "new_safe_range_begin" - this is
+		// because choosing any other version will invalidate the changes that we made to the peek logic
```

Comment:
Shall I also mention we may not copy the correct version range during version vector recovery if this index is not used?

Reply:
Yes. It's important to mention that, since that is critical to the correctness.

Reply:
Done.